File: meta/revised/2024/DEC/13/x/x.py
class Solution:

    def mark(self, matrix, i, j):
        rows = len(matrix)
        cols = len(matrix[0])

        # mark the row
        for col in range(0, cols):
            if matrix[i][col]==-1 and j!=col:
                continue
            matrix[i][col] = 0

        # mark the col
        for row in range(0, rows):
            if matrix[row][j]==-1 and i!=row:
                continue
            matrix[row][j] = 0

    def setZeroes(self, matrix: List[List[int]]) -> None:
        
        # marks each row 0 -> -1
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if (matrix[i][j]==0):
                    matrix[i][j]=-1
        
        # marks every cross as 0
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if (matrix[i][j]==-1):
                    self.mark(matrix, i, j)
        

File: meta/revised/2024/DEC/14/rotate matrix/x.ts

/**
 * Right => Transpose + reverse col list of ever row
 * Left => Transpose + reverse the row list
 */
export function rotateMatrix(mat: Array<Array<number>>){
    
    let rows = mat.length
    let cols = mat[0].length

    // transpose the matrix
    for (let i=0; i<rows; i++){
        for (let j=i+1; j<cols; j++){
                let temp = mat[i][j]
                mat[i][j]=mat[j][i]
                mat[j][i]=temp
        }
    }

    // reversing every row
    for (let row of mat){
        row.reverse()
    }

}

const mat = [
    [1,2,3],
    [4,5,6],
    [7,8,9]
]
rotateMatrix(mat)
console.log(mat)